Add PgBindIter for encoding and use it as the implementation encoding &[T]

tylerhawkes opened the pull request:
Partial fix for postgres of https://github.com/launchbadge/sqlx/issues/294

I've been using this implementation in several different projects and thought that I would see if it can be accepted upstream. I don't care about the name, but it is only useful when calling `.bind(PgBindIter(iter...))`, so that's what I've been calling it.

I use this for writing inserts using unnest in postgres and it can easily hit 100k rows/sec on any hardware I've run it on (my mac, ec2, gcp, etc.)

The only overhead this implementation has over the current one for `&[T]` is the `count += 1` in the encoding loop. Hopefully this gets optimized out for slices due to the check in the calling function before delegating the call.

The existing tests for `&[T]` should cover testing this since it is now the implementation.

I have used a peeking iterator for the first stuff in the past, but that probably has some iteration overhead that is larger than just handling the first item separately.

tylerhawkes committed:
Add PgBindIter for encoding and use it as the implementation encoding &[T]

paolobarbolini commented on sqlx-postgres/src/bind_iter.rs:
```diff
@@ -0,0 +1,114 @@
+        // need ownership to iterate
+        mut iter: I,
+        buf: &mut PgArgumentBuffer,
+    ) -> Result<IsNull, Box<dyn std::error::Error + Send + Sync + 'static>> {
```
```suggestion
    ) -> Result<IsNull, BoxDynError> {
```

paolobarbolini commented on sqlx-postgres/src/bind_iter.rs:
```diff
@@ -0,0 +1,114 @@
+    fn encode_by_ref(
+        &self,
+        buf: &mut PgArgumentBuffer,
+    ) -> Result<IsNull, Box<dyn std::error::Error + Send + Sync + 'static>> {
```
```suggestion
    ) -> Result<IsNull, BoxDynError> {
```

paolobarbolini commented on sqlx-postgres/src/bind_iter.rs:
```diff
@@ -0,0 +1,114 @@
+    fn encode(
+        self,
+        buf: &mut PgArgumentBuffer,
+    ) -> Result<IsNull, Box<dyn std::error::Error + Send + Sync + 'static>>
```
```suggestion
    ) -> Result<IsNull, BoxDynError>
```

paolobarbolini commented on sqlx-postgres/src/bind_iter.rs:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        // set the length now that we know what it is.
+        buf[len_start..(len_start + 4)].copy_from_slice(count.to_be_bytes().as_slice());
```
```suggestion
        buf[len_start..(len_start + 4)].copy_from_slice(&count.to_be_bytes());
```

paolobarbolini commented on sqlx-postgres/src/bind_iter.rs:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        let mut count = 1_i32;
+        const MAX: usize = i32::MAX as usize;
```
This looks like an oversight. We've already written 1 item at this point.

```suggestion
        const MAX: usize = i32::MAX as usize - 1;
```

tylerhawkes replied:
Good catch. I missed this when removing the peeking iterator.

paolobarbolini commented on sqlx-postgres/src/bind_iter.rs:
```diff
@@ -0,0 +1,114 @@
+
+        const OVERFLOW: usize = MAX + 1;
+        if iter.next().is_some() {
+            return Err(format!("encoded iterator is too large for Postgres: {OVERFLOW}").into());
```
I'm not sure if printing the `Iterator::size_hint` would be more interesting here

tylerhawkes replied:
I've got it printing the larger of the `size_hint` lower bound and `OVERFLOW` now

paolobarbolini reviewed (commented):
Drive-by review: I think this is a good idea. I've left a couple of mostly nitpicks.

tylerhawkes committed:
Implement suggestions from review

tylerhawkes committed:
Add docs to PgBindIter and test to ensure it works for owned and borrowed types

tylerhawkes commented:
I added a few lines of documentation about why this is useful and how to best use it in postgres with the unnest operator. I also added a test that ensures it works as expected and is usable for owned and borrowed types returned in the iterator.

tylerhawkes commented:
Is there anything else I need to do to get this merged?

tylerhawkes committed:
Use extension trait for iterators to allow code to flow better. Make struct private. Don't reference unneeded generic T. Make doc tests compile.

tylerhawkes commented:
I thought about it more and using an extension trait feels a lot nicer of an api to use and it lets us hide some implementation details that aren't important.

Instead of `bind_iter` as the fn name, maybe `into_bind` would work better or some other name. This feels nicer to write at the very least.

tylerhawkes committed:
Fix doc function

tylerhawkes committed:
Fix doc test to actually compile

abonander commented on sqlx-postgres/src/bind_iter.rs:
```diff
@@ -0,0 +1,152 @@
+
+impl<'q, I> Encode<'q, Postgres> for PgBindIter<I>
+where
+    // Clone is required for the encode_by_ref call since we can't iterate with a shared reference
+    I: Iterator + Clone,
```
The plan I had for this was to use `Cell<Option<I>>` since the `Clone` bound could theoretically be hard to satisfy for some iterators.

Using `Cell` would make it `!Sync` and only able to used once, but those seem like pretty reasonable tradeoffs. I don't expect anyone to need to share this across threads or or need to bind the same iterator more than once (they could just use the same placeholder number to refer to the existing binding or create a new iterator).

It seems like maybe we should split `Encode` into two traits: one that encodes by-value and one by-reference, then a blanket impl of the former over the latter. The by-reference trait would be the one that most types would implement, but the by-value trait would be the one that all the APIs actually accept.

tylerhawkes replied:
Ok, I can make the change. I don't think it will affect the send-ness of the execution future so that should be fine.

I agree with splitting up the Encode trait. It would make this simpler.

tylerhawkes replied:
@abonander I've made the change to `Cell<Option<I>>`

tylerhawkes committed:
Use Cell<Option<I>> instead of Clone bound

tylerhawkes committed:
Merge remote-tracking branch 'upstream/main' into add-postgres-bind-iter

tylerhawkes commented:
@abonander Can we get this included or something like it for 0.9?

cheban96 commented:
Hi, intersting it's will be working with macros?
